crux_cli/src/codegen/mod.rs

```diff
@@ -0,0 +1,160 @@
+mod filter;
+mod formatter;
+pub mod generate;
+mod indexed;
+mod item;
+mod node;
+mod serde;
+mod serde_generate;
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    fs::{self, File},
+    io::Read,
+    path::PathBuf,
+    process::Command,
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+use guppy::{graph::PackageGraph, MetadataCommand};
+use log::debug;
+use rustdoc_types::Crate;
+
+use crate::args::CodegenArgs;
+use filter::Filter;
+use formatter::Formatter;
+use node::ItemNode;
+use serde_generate::format::ContainerFormat;
+
+pub type Registry = BTreeMap<String, ContainerFormat>;
+
+pub fn codegen(args: &CodegenArgs) -> Result<()> {
+    let mut cmd = MetadataCommand::new();
+    let package_graph = PackageGraph::from_command(&mut cmd)?;
+
+    let manifest_paths: BTreeMap<&str, &str> = package_graph
+        .packages()
+        .map(|package| (package.name(), package.manifest_path().as_str()))
+        .collect();
+
+    let Ok(lib) = package_graph.workspace().member_by_path(&args.lib) else {
+        bail!("Could not find workspace package with path {}", args.lib)
+    };
+
+    let lib_name = lib.name();
+
+    let registry = run(lib_name, |name| load_crate(name, &manifest_paths))?;
+
+    // switch from vendored types to `serde-reflection` types
+    let registry: serde_reflection::Registry =
+        serde_json::from_slice(&serde_json::to_vec(&registry)?)?;
```

Comment:
What's the plan for this in the long run? If any incompatibility happens there, the CLI will just blow up, right?

Reply:
Yeah. I used the types from their lib as the intermediate representation, but had to vendor them to add things like `impl Hash` for `ascent`. Then wanted to use the published types for generation, hence the ser/de.
I think in the long run we want to use our own types throughout, and obvs control the generation ourselves. This is just an intermediate step (that also ensures compatibility with what we are doing today).